packages/frontend-internal/src/wiring/OpaqueType.ts

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// TODO(Rugvip): This lives here temporarily, but should be moved to a more
+// central location. It's useful for backend packages too so we'll need to have
+// it in a common package, but it might also be that we want to make it
+// available publicly too in which case it would make sense to have this be part
+// of @backstage/version-bridge. The problem with exporting it from there is
+// that it would need to be very stable at that point, so it might be a bit
+// early to put it there already.
+
+/**
+ * A helper for working with opaque types.
+ */
+export class OpaqueType<
+  T extends {
+    public: { $$type: string };
+    versions: { version: string | undefined };
+  },
+> {
+  /**
+   * Creates a new opaque type.
+   *
+   * @param options.type The type identifier of the opaque type
+   * @param options.versions The available versions of the opaque type
+   * @returns A new opaque type helper
+   */
+  static create<
+    T extends {
+      public: { $$type: string };
+      versions: { version: string | undefined };
+    },
+  >(options: {
+    type: T['public']['$$type'];
+    versions: Array<T['versions']['version']>;
+  }) {
+    return new OpaqueType<T>(options.type, new Set(options.versions));
+  }
+
+  #type: string;
+  #versions: Set<string | undefined>;
```

Comment:
why `| undefined`?

Reply:
There are quite a couple of existing opaque types that don't define a particular version, and I want to be able to use this for those types as well. This just lets you explicitly say that `undefined` is a supported version.